src/lib.rs

```diff
@@ -224,13 +225,70 @@ fn fd_cloexec(fd: u32) -> io::Result<()> {
     Ok(())
 }
 
+/// Asks the service manager for enabled watchdog.
+///
+/// If the `unset_env` parameter is set, the `WATCHDOG_USEC` and `WATCHDOG_PID` environment variables
+/// will be unset before returning. Further calls to `watchdog_enabled` will fail, but
+/// child processes will no longer inherit the variable.
+///
+/// See [`sd_watchdog_enabled(3)`][sd_watchdog_enabled] for details.
+///
+/// [sd_watchdog_enabled]: https://www.freedesktop.org/software/systemd/man/sd_watchdog_enabled.html
+///
+///
+/// # Example
+///
+/// ```no_run
+/// # use sd_notify;
+/// #
+/// let mut usec = 0;
+/// let enabled = sd_notify::watchdog_enabled(true, &mut usec);
+/// ```
+pub fn watchdog_enabled(unset_env: bool, usec: &mut u64) -> io::Result<bool> {
+    struct Guard {
+        unset_env: bool,
+    }
+
+    impl Drop for Guard {
+        fn drop(&mut self) {
+            if self.unset_env {
+                env::remove_var("WATCHDOG_USEC");
+                env::remove_var("WATCHDOG_PID");
+            }
+        }
+    }
+
+    let _guard = Guard { unset_env };
+
+    let s = env::var_os("WATCHDOG_USEC");
+    let p = env::var_os("WATCHDOG_PID");
+
+    if let Some(t) = s
+        .to_owned()
+        .and_then(|s| s.to_str().and_then(|s| u64::from_str(s).ok()))
+    {
+        if let Some(pid) = p
+            .to_owned()
+            .and_then(|s| s.to_str().and_then(|s| u32::from_str(s).ok()))
+        {
+            if process::id() == pid {
```

Comment:
This doesn't look right. The docs say:

> If the `$WATCHDOG_USEC` environment variable is set, and the `$WATCHDOG_PID` variable is unset or set to the PID of the current process, the service manager expects notifications from this process.

So if `$WATCHDOG_PID` is unset and `$WATCHDOG_USEC` is set, the watchdog is actually enabled, so we should return `true`.

I'd also be inclined to drop the `Result`, or to return `Err` if parsing the two values fails.

Reply:
You might be right. I didn't look on the spec so much. Instead I tried to implement the same logic provided by the original [code](https://github.com/systemd/systemd/blob/f3376ee8fa28aab3f7edfad1ddfbcceca5bc841c/src/libsystemd/sd-daemon/sd-daemon.c#L632):
```
_public_ int sd_watchdog_enabled(int unset_environment, uint64_t *usec) {
        const char *s, *p = ""; /* p is set to dummy value to do unsetting */
        uint64_t u;
        int r = 0;

        s = getenv("WATCHDOG_USEC");
        if (!s)
                goto finish;

        r = safe_atou64(s, &u);
        if (r < 0)
                goto finish;
        if (u <= 0 || u >= USEC_INFINITY) {
                r = -EINVAL;
                goto finish;
        }

        p = getenv("WATCHDOG_PID");
        if (p) {
                pid_t pid;

                r = parse_pid(p, &pid);
                if (r < 0)
                        goto finish;

                /* Is this for us? */
                if (getpid_cached() != pid) {
                        r = 0;
                        goto finish;
                }
        }

        if (usec)
                *usec = u;

        r = 1;

finish:
        if (unset_environment && s)
                assert_se(unsetenv("WATCHDOG_USEC") == 0);
        if (unset_environment && p)
                assert_se(unsetenv("WATCHDOG_PID") == 0);

        return r;
}
```